### ISSUE TITLE
The Session Start scan certifies frontmatter it never parses — and the documented template is itself invalid YAML

### ISSUE THREAD

**Issue by Nikhiltogadiya** [rebelytics/one-skill-to-rule-them-all#81]
## What happens

SKILL.md's Session Start Protocol, step 2 (Scan) extracts the block between the first
two `---` lines with `awk` and treats a non-empty result as success:

```bash
hdr=$(awk 'NR==1 && /^---[[:space:]]*$/ {fm=1; next}
           fm && /^---[[:space:]]*$/ {exit}
           fm' "$f")
[ -n "$hdr" ] && parsed=$(( parsed + 1 ))
```

The block is never parsed. A file whose frontmatter is not valid YAML is therefore
counted as healthy, and the fields the per-file format exists to expose are unreadable
to anything that actually parses them. The same snippet appears again in
`references/observation-log.md` § Scanning cheaply.

In one first-run backfill, **19 of 22 observation files carried frontmatter that
`yaml.safe_load` refuses**, while the documented scan reported 22 files and 22 headers
parsed.

## Reproduction

Run the snippet verbatim against a file with an unquoted colon in a value:

```console
$ printf -- '---\narea: broken: unquoted colon\n---\nbody\n' > observation-log/0001-x.md
$ # SKILL.md step 2 snippet, unmodified
1 files, 1 headers parsed          # reports healthy

$ python3 -c "import yaml; yaml.safe_load(open('observation-log/0001-x.md').read().split('---\n',2)[1])"
yaml.scanner.ScannerError: mapping values are not allowed here
```

Both commands were run from a clean shell on Linux, GNU awk, PyYAML 6.0.1.

## The documented template does not parse either

SKILL.md § How to Log ("File format") publishes the frontmatter template. Parsed
exactly as published:

```
RESULT: DOES NOT PARSE
error: while parsing a flow sequence, line 9, column 19:
       siblings_checked: [MANDATORY, never blank: the fam ...
```

Three lines fail individually — `skill:`, `proposes_skill:` and `siblings_checked:` —
because the bracketed guidance contains `: ` inside a flow sequence.

Those brackets are placeholders meant to be replaced, so the template itself is not
the artefact. The failure survives the substitution, which is the point: the fields
that carry prose — `area`, `session_context`, `siblings_checked`, `resolution`, and
any `title` containing a colon — become invalid YAML whenever that prose contains
`: `. For these fields that is the common case, not the exotic one. All 19 broken
files above broke exactly that way, written by following the template.

## Why the existing guard cannot catch it

Step 2 already carries a guard:

> `[ "$n" -gt 0 ] && [ "$parsed" -eq 0 ] && { echo "SCAN COMMAND BROKEN — ..."; exit 1; }`

It tests for an **empty** result. `awk` extracts the text of an invalid header without
difficulty, so `parsed` equals `n` and the guard stays silent. The check certifies that
a block was *extracted*; the property the format depends on is that it can be *read*.

This is load-bearing rather than cosmetic. `references/observation-log.md` § Layout
calls the frontmatter "the metadata every scan reads", and § Scanning cheaply
justifies the per-file layout by the scan staying cheap "once hundreds of observations
exist". A review reading `status`, `skill` and `proposes_skill` through any YAML
parser would have thrown on 19 of those 22 files, with nothing earlier reporting a
problem.

## Suggested fix

Two parts, either useful alone.

**1. Make the scan's success condition parsing, not extraction.** Where a YAML parser
is available, load each header and count the failures, then report that count beside
the open-observation count so a corrupt log announces itself at session start rather
than at the first review that needs the fields. Where no parser is available, at
minimum reject a header whose lines do not all match `key:` followed by a value
containing no unquoted `: `.

The parser version was implemented here as a session-start hook in about fifteen
lines, including the optional-import fallback, and verified against fixtures for both
the firing and the correctly-silent case. A pull request can follow if that would
help.

**2. Quote the prose-bearing fields in the template.** Show `area`,
`session_context`, `siblings_checked`, `resolution` and `title` double-quoted in the
File format block, and make the `skill:` / `proposes_skill:` examples quoted flow
lists — `skill: ["a", "b"]`. A writing session copies the template's shape, so the
template is the cheapest place to fix this.

One knock-on: `skill:` entries for plugin-scoped skills contain a colon
(`superpowers:verification-before-completion`). Unquoted inside a flow sequence those
parse as nested mappings rather than strings even when the file is otherwise valid, so
the quoting in part 2 also closes a second, quieter failure.

## Related

Adjacent but distinct. **#67** (post-write verification only handles the `>1` case; a
count of `0` passes silently) is the same shape — a mandatory check whose success
condition tests the wrong property, so the log is wrong while every check reports
success — but concerns entry numbering in the pre-3.0 single-file log, where this
concerns frontmatter validity in the per-file log that replaced it. **#28** and
**#61** also concern a validator, but `scripts/validate-skill-bundle.py` for skill
delivery, not the observation-log scan.

## Environment

Installed from the `v3.0.0` release bundle. Verified present at `main` @ `510caad2`:
the snippet appears in `SKILL.md` (Session Start Protocol, step 2) and in
`references/observation-log.md` (§ Scanning cheaply), and the template block in
SKILL.md § How to Log reproduces the parse failure above.

**Comment by notbucki** [comment]
A second measurement of this from a different corpus, plus two increments the report does not
cover yet.

**Confirmation.** Same snippet, different log: **15 of 55 files** (6 active, 9 archived) failed
`yaml.safe_load` while the step-2 scan reported **28 of 28 headers** extracted. Same shape as
your 19-of-22, on a log written by a different person against the same template.

**Increment 1 — a second failure class, and the parser-less fallback does not catch it.**
Besides the unquoted `: ` you reproduce, prose breaks the header a second way: a typographic
opening quote (`„`, `“`) closed with an ASCII `"` ends a double-quoted scalar early. The pair
looks correct in an editor, and it is the natural thing to type when a title quotes something.

Run from a clean shell (GNU awk 5.2.1, PyYAML 6.0.1):

```console
$ printf -- '---\ntitle: "Er sagte „ja" und ging"\nstatus: open\n---\nbody\n' \
    > observation-log/0001-quote.md
$ # SKILL.md step 2 snippet, unmodified
2 files, 2 headers parsed          # reports healthy

$ python3 -c "import yaml,io;yaml.safe_load(io.open('observation-log/0001-quote.md',encoding='utf-8').read().split('---')[1])"
yaml.scanner.ScannerError: while parsing a block mapping
```

Why it matters for part 1 of your fix: that file contains **no** unquoted `: `, so the
parser-less fallback ("at minimum reject a header whose lines do not all match …") passes it.
Where a parser is available the class is caught for free; where one is not, the fallback needs
to name both classes, or it certifies exactly the file that started this.

**Increment 2 — the validation misses `archive/`.** The snippet's glob is
`observation-log/*.md`, and `archive/` is a subdirectory, so a check written against that glob
reports on part of the corpus and the answer looks complete. Measured on the log above: the
first audit reported 6 broken files, the actual number was 15. Same three files, both globs:

```console
$ # glob as written in the snippet
checked 2, invalid: 1
$ # same check, plus observation-log/archive/*.md
checked 3, invalid: 2
```

Two small things that follow from it: whatever the check becomes should glob `archive/` as
well, and "no YAML parser available" is worth reporting as its own outcome — a validation step
that silently degrades to "fine" is the same failure one level up.

Both increments exist as a tested patch on top of your part 1 and part 2 if that is useful;
otherwise this is just the extra data, and the fix design in this issue already covers the
main case.

---

*Contributed via Claude Code (Anthropic). The agent reproduced the failure, searched existing
issues and pull requests, and drafted this comment; a human reviewed and approved it before
posting.*

**Comment by rebelytics** [comment]
This is fixed on `release/v3.2.0`, the branch that goes live as v3.2.0 next week after a week of local testing — the template quotes every prose value, and the scan reports a suspect-header count beside the file count. Your regex idea measured well on my own log: 292 files, 27 failing yaml.safe_load, 26 caught with zero false positives — which is how I found I had 27 of my own to fix. If you can run the branch against your setup this week, that confirmation is what carries it into main. Thanks for the report.

**Comment by Nikhiltogadiya** [comment]
Thanks for the update and for addressing this. Yes, I can run the `release/v3.2.0` branch against my setup and verify the fix. I will report back with the results once I have tested it.

**Comment by notbucki** [comment]
Ran `release/v3.2.0` @ 8112ff3 against this setup as a second data point next to @Nikhiltogadiya's.

**Confirmed.** The template parses as published (`yaml.safe_load`, 14 keys). The issue's original repro and the `resolution: Actioned: …` shape are both flagged. Three local logs, 322 files including `archive/`: no file with an unquoted `: ` left, and zero false positives from the new pattern on the 313 that parse.

**Not covered — the two increments from the comment above are still open.** The check as shipped flagged 0 of the 6 headers in that corpus that still fail `yaml.safe_load`:

- 6 of 6 start the value with `"` and close it early (3 inner ASCII `"`, 2 typographic `„…"` pairs, 1 backtick). The pattern's `[^"'[|>]` class excludes every one of them by construction.
- 6 of 6 sit in `archive/`. Every `find` in the snippet is `-maxdepth 1`, so for that log the scan printed `files: 1  parsed: 1  suspect: 0` and no NOTE.

Minimal repro for the class, file in the active directory (GNU awk 5.2.1, PyYAML 6.0.1):

```console
$ printf -- '---\ntitle: "Er sagte „ja" und ging"\nstatus: open\n---\nbody\n' > observation-log/0001-x.md
$ # SKILL.md step 2 snippet @ 8112ff3, unmodified
files: 1  parsed: 1  suspect: 0
$ python3 -c "import yaml,io;yaml.safe_load(io.open('observation-log/0001-x.md',encoding='utf-8').read().split('---')[1])"
yaml.parser.ParserError: while parsing a block mapping
```

Two more awk rules in the same snippet catch all 6 here with 0 false positives on the 322 files:

```awk
fm && /^[a-z_]+: "([^"\\]|\\.)*"[[:space:]]*[^[:space:]#]/ {print FILENAME; nextfile}   # text after the closing quote
fm && /^[a-z_]+: [`@%]/ {print FILENAME; nextfile}                                   # value starts with a char no token can start with
```

For `archive/`, a second count without `-maxdepth 1` on the same `printf` line would make the number cover the corpus the reviews actually read. Happy to send both as a small PR against `release/v3.2.0` if that helps.

**One false positive in the fix itself.** The template's own line `reference:              # optional: path to saved session-local evidence` matches the pattern (the `: ` is inside the trailing comment), so a file that keeps the template's comments trips the NOTE at every session start although it parses. Dropping that colon from the comment fixes it.

**Correction to the knock-on in the issue body.** `skill: [superpowers:verification-before-completion]` loads as a plain string on PyYAML ≥ 5.1 (colon in a plain scalar in flow context, pyyaml PR #45); only `[a: b]` with a space nests. 26 such entries in these logs, all strings — the bare-kebab-case advice in the new File format paragraph is safe.

---

*Drafted with Claude Code (Claude Fable 5.1): the agent ran the branch's scan and a YAML parser over three local logs and a fixture set and drafted this comment; a human reviewed and approved it before posting.*

**Comment by notbucki** [comment]
Two corrections to my comment above, and three things an independent review of the proposed rules turned up before anyone copies them.

**Corrections.** (1) 5 of the 6 unreadable headers start the value with `"` and close it early; the sixth starts with a backtick — not "6 of 6 start with `"`". (2) `skill: [superpowers:verification-before-completion]` loads as a list containing one string. Nesting into a mapping happens when the colon is followed by a space or ends the scalar (`[a: b]`, `[a:]`, `["a":b]`), not only with a space — the point stands: bare `plugin:skill` names are safe in the new template.

**The rules as posted require exactly one space after the colon.** `title:  "Er sagte „ja" und ging"` (two spaces) fails `yaml.safe_load` and passes all of them. The same spacing turns the shipped rule into a false positive: `title:  "Valid: padded"` is valid YAML and is counted as suspect, because the second space is the "first value character". Both go away with `:[ ]+` and a space-excluding class.

**The shipped awk carries a state leak across files.** A match ends with `nextfile` while `fm` is still 1; if the next file in the batch has no opening `---`, its body is scanned with the previous file's state. Reproduced with GNU awk 5.2.1 and mawk 1.3.4: a flagged header followed by a legacy log containing `area: x: y` in its body reports two suspects, the legacy file alone reports none. `archive/` is where such legacy files live, so the count becomes file-order dependent exactly where the second increment adds it. A per-file reset on `FNR==1` closes it.

Corrected program, one string used for both directories (6 of 6 caught, 0 false positives on 323 files, gawk and mawk identical; an absent `archive/` prints `archive-suspect: 0` under `bash -e -o pipefail`):

```awk
FNR==1 {fm = (/^---[[:space:]]*$/ ? 1 : 0); if (!fm) nextfile; next}
  fm && /^---[[:space:]]*$/ {fm=0; nextfile}
  fm && /^[a-z_]+:[ ]+[^"\047[{|>#[:space:]].*: / {print FILENAME; nextfile}
  fm && /^[a-z_]+:[ ]+"([^"\\]|\\.)*"[[:space:]]*[^[:space:]#]/ {print FILENAME; nextfile}
  fm && /^[a-z_]+:[ ]+[`@%]/ {print FILENAME; nextfile}
  fm && /^[a-z_]+:[ ]+\{[^,{"\047}]*: [^,{"\047}]*: / {print FILENAME; nextfile}
```

**Known limits, so the count stays a lower bound:** an unterminated quote, a tab in the value, a colon at end of line, a key outside `[a-z_]+`, `[a: b]` inside a flow list (valid YAML, silently a mapping), a malformed flow list, and a malformed flow map beyond its first entry (`{a: b, c: d: e}` — the last rule checks only the first entry of a flat, unquoted map) are not reported; a value followed by a comment that contains `: ` (`status: open   # note: x`) is reported although it parses.

---

*Drafted with Claude Code (Claude Fable 5.1). A read-only Codex adversarial review (gpt-6-astra, reasoning effort max) of the patch and of the comment above found the spacing gap, the state leak and the absent-archive failure, and flagged the two wording errors; the corrections were reproduced and fixed before posting. A human reviewed and approved this comment.*

*Edited 2026-09-05: three further Codex adversarial review rounds (gpt-6-astra, max) over the fixes found that the unquoted-colon rule — as shipped in v3.2.0 and as first posted here — flags a valid flow map such as `skill_qualifiers: {task-observer: "Session Start"}` (its class excludes `[` but not `{`), and that simply excluding `{` would hide a malformed one such as `{task-observer: Session Start: Scan}`. The program above now excludes `{` from that rule and adds a last rule for a flat, unquoted map whose first entry carries a second `: `; a malformed later entry is a documented limit (see above). While testing that rule end to end, one more trap surfaced: a literal `{}` inside the awk program (as in a bracket class `[^,{}"]`) is replaced by `find -exec … {} +`, so the snippet silently reports 0 while a direct `awk` call reports hits — the classes above are written to avoid it. The 323-file numbers are unchanged.*
